Source/WebCore/Modules/url-pattern/URLPatternParser.h

```diff
@@ -0,0 +1,90 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+namespace WebCore {
+namespace URLPatternUtilities {
+
+struct Token;
+enum class TokenType : uint8_t;
+
+enum class PartType : uint8_t { FixedText, Regexp, SegmentWildcard, FullWildcard };
+enum class Modifier : uint8_t { None, Optional, ZeroOrMore, OneOrMore };
+
+struct Part {
+    PartType type;
+    String value;
+    Modifier modifier;
+    String name { };
+    String prefix { };
+    String suffix { };
+};
+
+struct URLPatternStringOptions {
+    String delimiterCodepoint { emptyString() };
+    String prefixCodepoint { };
+    bool ignoreCase { false };
+};
+
+class URLPatternParser {
+public:
+    URLPatternParser(EncodingCallbackType, String&& segmentWildcardRegexp);
+    ExceptionOr<void> performParse(const URLPatternStringOptions& patternSettings);
+
+    void setTokenList(Vector<Token>&& tokenList) { m_tokenList = tokenList; }
+    Vector<Part> getPartList(void) { return m_partList; }
+
+private:
+    Token tryToConsumeToken(TokenType consumedType);
+    Token tryToConsumeRegexOrWildcardToken(const Token& nameToken);
+    Token tryToConsumeModifierToken();
+
+    String consumeText();
+    ExceptionOr<Token> consumeRequiredToken(TokenType requiredType);
+
+    ExceptionOr<void> maybeAddPartFromPendingFixedValue();
+    ExceptionOr<void> addPart(String&& prefix, const Token& nameToken, const Token& regexpOrWildcardToken, String&& suffix, const Token& modifierToken);
+
+    bool isDuplicateName(StringView name) const;
+
+    Vector<Token> m_tokenList;
+    Vector<Part> m_partList;
+    EncodingCallbackType m_callbackType;
+    String m_segmentWildcardRegexp;
+    String m_pendingFixedValue { };
+    size_t m_index { 0 };
+    int m_nextNumericName { 0 };
+};
+
+ExceptionOr<Vector<Part>> parse(StringView patternStringInput, const URLPatternStringOptions& patternSettings,  EncodingCallbackType);
+
+// FIXME: Consider moving functions to somewhere generic, perhaps refactor Part to its own class.
+String generateSegmentWildcardRegexp(const URLPatternStringOptions& patternStringSettings);
```

Comment:
I think we can omit `patternStringSettings`. It doesn't bring much info given the parameter type.